src/compiler/factory.ts

```diff
@@ -122,4 +122,210 @@ namespace ts {
+        node.condition = condition;
+        node.questionToken = createSynthesizedNode(SyntaxKind.QualifiedName);
+        node.whenTrue = whenTrue;
+        node.colonToken = createSynthesizedNode(SyntaxKind.ColonToken);
+        node.whenFalse = whenFalse;
+        return node;
+    }
+
+    export function createBinary(left: Expression, operator: SyntaxKind, right: Expression, location?: TextRange) {
+        const node = <BinaryExpression>createNode(SyntaxKind.BinaryExpression, location);
+        node.left = parenthesizeForBinary(left, operator, BinaryOperand.Left);
+        node.operatorToken = createSynthesizedNode(operator);
+        node.right = parenthesizeForBinary(right, operator, BinaryOperand.Right);
+        return node;
+    }
+
+    export function createAssignment(left: Expression, right: Expression, location?: TextRange) {
+        return createBinary(left, SyntaxKind.EqualsToken, right, location);
+    }
+
+    export function createStrictEquality(left: Expression, right: Expression) {
+        return createBinary(left, SyntaxKind.EqualsEqualsEqualsToken, right);
+    }
+
+    export function createComma(left: Expression, right: Expression) {
+        return <Expression>createBinary(left, SyntaxKind.CommaToken, right);
+    }
+
+    export function createCall(expression: Expression, argumentsArray: Expression[]) {
+        const node = <CallExpression>createNode(SyntaxKind.CallExpression);
+        node.expression = parenthesizeForAccess(expression);
+        node.arguments = createNodeArray(argumentsArray);
+        return node;
+    }
+
+    export function createArraySlice(array: Expression, start?: number | Expression) {
+        const argumentsList: Expression[] = start !== undefined ? [coerceExpression(start)] : [];
+        return createCall(createPropertyAccess(array, "slice"), argumentsList);
+    }
+
+    export function parenthesizeExpression(expression: Expression) {
+        const node = <ParenthesizedExpression>createNode(SyntaxKind.ParenthesizedExpression);
+        node.expression = expression;
+        return node;
+    }
+
+    export function inlineExpressions(expressions: Expression[]) {
+        return reduceLeft(expressions, createComma);
+    }
+
+    function coerceIdentifier(value: string | Identifier) {
+        if (typeof value === "string") {
+            return createIdentifier(value);
+        }
+        else {
+            return value;
+        }
+    }
+
+    function coerceExpression(value: string | number | boolean | Expression): Expression {
+        if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
+            return createLiteral(value);
+        }
+        else {
+            return value;
+        }
+    }
+
+    const enum BinaryOperand {
+        Left,
+        Right
+    }
+
+    function parenthesizeForBinary(operand: Expression, operator: SyntaxKind, side: BinaryOperand) {

```

Comment:
It sounds more like this is `parenthesizeForBinaryOperand`. Can you make this this `binaryOperator`, `operand`, `operandSide` in that order?